### PR TITLE
fixed resolve mutates input values

### PR DESCRIPTION
**Pull request by roironn**
_Fixed resolve mutates input values_

---

_The resolve method should not modify its input values. The previous implementation mutated the input in two place - fields, annotation types, and refType). This PR fixes that._

---
_Release Notes_: 
_NA_

### CODE REVIEW

File: packages/workspace/test/core/expressions.test.ts [hunk 1]
```diff
@@ -164,6 +164,8 @@ describe('Test Salto Expressions', () => {
       simpleRefType,
     ]
 
+    const origElements = _.cloneDeep(elements)
```
Comment (ori-moisis) [inline packages/workspace/test/core/expressions.test.ts:167]: we should never cloneDeep elements

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 2]
```diff
@@ -203,7 +203,7 @@ const createNaclFilesState = async (
   }),
   mergedElements: new RemoteElementSource(await remoteMapCreator<Element>({
     namespace: getRemoteMapNamespace('merged', sourceName),
-    serialize: element => serialize([element]),
+    serialize: element => serialize([element], 'keepRef'),
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:206]: unrelated change?

File: packages/workspace/src/expressions.ts [hunk 3]
```diff
@@ -260,18 +264,23 @@ export const resolve = async (
   elementsSource: ReadOnlyElementsSource,
   inPlace = false
 ): Promise<Element[]> => {
-  // intentionally shallow clone because in resolve element we replace only top level properties
```
Comment (ori-moisis) [inline packages/workspace/src/expressions.ts:263]: why did you remove the comment?
seems like we are still using shallow copy...

File: packages/workspace/src/expressions.ts [hunk 4]
```diff
@@ -42,24 +42,29 @@ export class CircularReference {
...
-      elementsSource,
-      workingSetElements,
-    )
+  if (workingSetElements[elemID.getFullName()] !== undefined) {
+    return workingSetElements[elemID.getFullName()].element
   }
-
-  return unresolvedElement
+  const element = await elementsSource.get(elemID)
+  if (element !== undefined && inPlace) {
```
Comment (ori-moisis) [inline packages/workspace/src/expressions.ts:64]: we create a copy if `inPlace` is `true`?
that seems... wrong...

also, why do we even have the option to resolve in place? in what scenario is that not a bug?

File: packages/workspace/src/expressions.ts [hunk 5]
```diff
@@ -42,24 +42,29 @@ export class CircularReference {
...
+    return shallowCloneElement((await elementsSource.get(elemID)))
```
Comment (ori-moisis) [inline packages/workspace/src/expressions.ts:65]: why get from the element source twice?
also, how do we use the result from this function? is a shallow clone enough here or is this going to modify an element from the **ReadOnly**ElementsSource?

File: packages/workspace/src/expressions.ts [hunk 6]
```diff
@@ -88,8 +93,12 @@ const resolveMaybeExpression: Resolver<Value> = async (
   // all we need to do is make it point to the element from the context. If the element is not in
   // the context or the source - we'll keep it as it is.
   if (isElement(value)) {
-    return (await getResolvedElement(value.elemID, elementsSource, workingSetElements))
-      ?? value
+    // eslint-disable-next-line no-use-before-define
```
Comment (ori-moisis) [inline packages/workspace/src/expressions.ts:96]: the comment above this line (in lines 91-94) says completely the opposite of what happens in the code

File: packages/workspace/src/expressions.ts [hunk 7]
```diff
@@ -171,11 +180,6 @@ const resolveElement = async (
   // to be used to resolve types. If it was already resolved use the reolsved and if not
   // fallback to the elementsSource
 
-  const contextedElementsGetter: ReadOnlyElementsSource = {
```
Comment (ori-moisis) [inline packages/workspace/src/expressions.ts:174]: I assume the comment above was referring to this code you just deleted or am I misreading?

**Review by ori-moisis — COMMENTED** [review]
there are a lot of changes in this logic that I don't understand (before this PR), would probably be best if someone who reviewed these changes would take a look at this PR

File: packages/workspace/src/expressions.ts (hunk 3 above)
Comment (roironn) [reply packages/workspace/src/expressions.ts:263]: I figured that the naming of the function is good enough - and since we also clone nested elements this might be misleading - I can re add it if you think it should have remained... thoughts?

File: packages/workspace/src/expressions.ts (hunk 4 above)
Comment (roironn) [reply packages/workspace/src/expressions.ts:64]: You are correct - we don't need this option here so I removed it.

Top - level in place modification is needed to resolve elements which are a part of the resulting plan graph - you can have a look at that usage - if you have a more elegant way to implement it I would be happy to remove this notion all together :)

**Review by adamro — APPROVED** [review]
